File: V3 - Deux clients et generation cles/client android/serverclient/app/src/main/java/com/example/server_client/Communication.java
/**
 * @file Communication.java
 * @brief High-level communication interface for Android messaging
 * @version 3.0
 * @author Alexis DEVERCHERE
 * @date 2025
 */

package com.example.server_client;
import android.util.Log;
import java.util.Arrays;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * @brief High-level communication manager for networked messaging
 * 
 * This class combines Connection and Protocol objects to provide a complete
 * messaging solution. It manages a background thread for continuous message
 * reception and provides methods for sending different types of commands.
 */
public class Communication {
    /** @brief Tag for Android logging */
    private static final String TAG = "Communication";
    
    /** @brief Buffer size for message reception */
    private static final int BUFFER_SIZE = 1024;

    /**
     * @brief Interface for handling received messages
     * 
     * Implementations receive callbacks when messages are decoded
     * from the network stream.
     */
    public interface MessageHandler {
        /**
         * @brief Called when a message is received and decoded
         * @param cmd The decoded command string
         * @param param The decoded parameter string
         */
        void onMessageReceived(String cmd, String param);
    }

    /** @brief Network connection object */
    private Connection connection;
    
    /** @brief Message protocol handler */
    private Protocol protocol;
    
    /** @brief Current message handler for callbacks */
    private MessageHandler messageHandler;
    
    /** @brief Background thread for message reception */
    private Thread communicationThread;
    
    /** @brief Thread-safe running flag */
    private AtomicBoolean running = new AtomicBoolean(false);

    /**
     * @brief Constructor for Communication class
     * 
     * Initializes the communication manager with the provided connection
     * and protocol objects. The communication thread is not started until
     * run() is called.
     * 
     * @param connection Network connection object
     * @param protocol Message protocol handler
     * 
     * @note The connection should be established before starting communication
     * @note The protocol object defines the message encoding/decoding format
     */
    public Communication(Connection connection, Protocol protocol) {
        this.connection = connection;
        this.protocol = protocol;
    }

    /**
     * @brief Set the message handler for incoming messages
     * 
     * Registers a callback interface that will be invoked when messages
     * are received and successfully decoded.
     * 
     * @param handler The MessageHandler implementation to receive callbacks
     * 
     * @note Only one handler can be active at a time
     * @note The handler is called from the communication thread context
     */
    public void setMessageHandler(MessageHandler handler) {
        this.messageHandler = handler;
    }

    /**
     * @brief Send a CMD_X message with parameter
     * 
     * Encodes and sends a CMD_X command message with the specified parameter.
     * This command type is typically used for client-to-server messaging.
     * 
     * @param param Parameter string to send with the command
     * 
     * @note The message is encoded using the protocol's encodeMessage method
     * @note Requires an active connection to succeed
     */
    public void comX(String param) {
        byte[] message = protocol.encodeMessage(Protocol.CMD_X, param);
        connection.write(message, message.length);
    }

    /**
     * @brief Send a CMD_Y message with parameter
     * 
     * Encodes and sends a CMD_Y command message with the specified parameter.
     * This command type is typically used for server-to-client messaging.
     * 
     * @param param Parameter string to send with the command
     * 
     * @note The message is encoded using the protocol's encodeMessage method
     * @note Requires an active connection to succeed
     * @note There appears to be a duplicate write call in the original code
     */
    public void comY(String param) {
        byte[] message = protocol.encodeMessage(Protocol.CMD_Y, param);
        connection.write(message, message.length);
        connection.write(message, message.length); // Note: Duplicate write in original
    }

    /**
     * @brief Start the communication thread for message reception
     * 
     * Creates and starts a background thread that continuously monitors
     * the connection for incoming messages. Messages are decoded and
     * passed to the registered message handler.
     * 
     * @note If already running, this method returns without creating a new thread
     * @note The thread includes connection health monitoring
     * @note Uses a 50ms delay between read attempts to prevent CPU hogging
     */
    public void run() {
        if (running.get()) {
            return; // Already running
        }

        running.set(true);

        communicationThread = new Thread(new Runnable() {
            @Override
            public void run() {
                Log.d(TAG, "Communication thread started");
                byte[] buffer = new byte[BUFFER_SIZE];
                int consecutiveEmptyReads = 0;
                final int maxConsecutiveEmptyReads = 5;

                while (running.get() && connection.isConnected()) {
                    Arrays.fill(buffer, (byte)0); // Clear buffer

                    int bytesRead = connection.read(buffer, BUFFER_SIZE - 1);

                    if (bytesRead > 0) {
                        consecutiveEmptyReads = 0;

                        // Ensure null termination and convert to string
                        String rawMessage = new String(buffer, 0, bytesRead);
                        Log.d(TAG, "Raw data received: '" + rawMessage + "'");

                        // Process the message
                        String[] parts = protocol.decodeMessage(rawMessage);

                        // Call the message handler if set
                        if (messageHandler != null) {
                            messageHandler.onMessageReceived(parts[0], parts[1]);
                        }
                    } else if (bytesRead == 0) {
                        // No data available
                        consecutiveEmptyReads++;

                        // If connection lost
                        if (!connection.isConnected()) {
                            Log.d(TAG, "Connection closed by peer (normal)");
                            break;
                        }

                        // Connection check after multiple empty reads
                        if (consecutiveEmptyReads > maxConsecutiveEmptyReads) {
                            Log.d(TAG, "Multiple empty reads, checking connection");
                            if (!connection.isConnected()) {
                                Log.d(TAG, "Connection appears broken");
                                break;
                            }
                            consecutiveEmptyReads = 0;
                        }
                    } else {
                        // Error occurred
                        Log.e(TAG, "Read error, terminating communication thread");
                        break;
                    }

                    // Small delay to prevent CPU hogging
                    try {
                        Thread.sleep(50);
                    } catch (InterruptedException e) {
                        break;
                    }
                }

                Log.d(TAG, "Communication thread exiting");
                running.set(false);
            }
        });

        communicationThread.start();
    }

    /**
     * @brief Stop the communication thread
     * 
     * Signals the communication thread to stop and waits for it to terminate.
     * This method blocks until the thread has completely stopped.
     * 
     * @note Safe to call multiple times
     * @note Waits up to 1 second for thread termination
     * @note If interrupted during join, a warning is logged
     */
    public void stop() {
        running.set(false);
        if (communicationThread != null) {
            communicationThread.interrupt();
            try {
                communicationThread.join(1000); // Wait up to 1 second for thread to finish
            } catch (InterruptedException e) {
                Log.w(TAG, "Interrupted while waiting for thread to join");
            }
            communicationThread = null;
        }
    }

    /**
     * @brief Check if the communication thread is currently running
     * 
     * @return boolean True if the communication thread is active, false otherwise
     * 
     * @note This reflects the thread state, not necessarily the connection state
     * @note Thread-safe using AtomicBoolean
     */
    public boolean isRunning() {
        return running.get();
    }
}

File: V3 - Deux clients et generation cles/serveur RaspberryPi/common/communication.c
/**
 * @file communication.c
 * @brief Implementation of high-level communication interface for networked messaging
 * @version 3.0
 * @author Alexis DEVERCHERE
 * @date 2025
 */

 #include "communication.h"
 #include <string.h>
 #include <unistd.h>
 #include <errno.h>
 
 /**
  * @brief Send a command X message with the specified parameter
  * 
  * Encodes a CMD_X message with the given parameter using the protocol
  * and sends it through the connection.
  * 
  * @param comm Pointer to the Communication object
  * @param param Parameter string to send with the command
  * 
  * @note The message is encoded using the protocol's encodeMessage function
  * @note The encoded message includes a null terminator in the transmission
  */
 static void Communication_comX(Communication* comm, const char* param) {
     char buffer[BUFFER_SIZE];
     comm->protocol->encodeMessage(comm->protocol, comm->protocol->cmdX, param, buffer);
     comm->connection->write(comm->connection, buffer, strlen(buffer) + 1);
 }
 
 /**
  * @brief Send a command Y message with the specified parameter
  * 
  * Encodes a CMD_Y message with the given parameter using the protocol
  * and sends it through the connection.
  * 
  * @param comm Pointer to the Communication object
  * @param param Parameter string to send with the command
  * 
  * @note The message is encoded using the protocol's encodeMessage function
  * @note The encoded message includes a null terminator in the transmission
  */
 static void Communication_comY(Communication* comm, const char* param) {
     char buffer[BUFFER_SIZE];
     comm->protocol->encodeMessage(comm->protocol, comm->protocol->cmdY, param, buffer);
     comm->connection->write(comm->connection, buffer, strlen(buffer) + 1);
 }
 
 /**
  * @brief Background thread function for continuous message reception
  * 
  * This function runs in a separate thread and continuously reads messages
  * from the connection, decodes them using the protocol, and calls the
  * message handler callback. It includes connection health monitoring
  * and graceful error handling.
  * 
  * @param arg Pointer to the Communication object (cast from void*)
  * @return void* Always returns NULL
  * 
  * @note The thread runs until comm->running is set to 0
  * @note Implements connection health monitoring with consecutive empty read detection
  * @note Uses a 50ms delay between read attempts to prevent CPU hogging
  * @note Performs connection probing after multiple consecutive empty reads
  */
 static void* communication_thread_function(void* arg) {
     Communication* comm = (Communication*)arg;
     char buffer[BUFFER_SIZE];
     char cmd[MAX_CMD_SIZE];
     char param[MAX_PARAM_SIZE];
     int consecutive_empty_reads = 0;
     const int max_consecutive_empty_reads = 5;
     
     printf("Communication thread started\n");
     
     while (comm->running) {
         // Clear buffers before each read
         memset(buffer, 0, BUFFER_SIZE);
         
         ssize_t bytes_read = comm->connection->read(comm->connection, buffer, BUFFER_SIZE - 1);
         
         if (bytes_read > 0) {
             // Reset empty read counter on successful data reception
             consecutive_empty_reads = 0;
             
             // Ensure null termination
             buffer[bytes_read] = '\0';
             printf("Raw data received: '%s'\n", buffer);
             
             // Clear command and parameter buffers
             memset(cmd, 0, MAX_CMD_SIZE);
             memset(param, 0, MAX_PARAM_SIZE);
             
             // Process the message
             comm->protocol->decodeMessage(comm->protocol, buffer, cmd, param);
             
             // Call the message handler if set
             if (comm->messageHandler) {
                 comm->messageHandler(cmd, param);
             }
         } else if (bytes_read == 0) {
             // No data available
             consecutive_empty_reads++;
             
             // Check if connection was marked as closed during read
             if (!comm->connection->connected) {
                 printf("Connection closed by peer (normal)\n");
                 break;
             }
             
             // If we've had too many consecutive empty reads, check connection
             if (consecutive_empty_reads > max_consecutive_empty_reads) {
                 // Try a zero-byte probe to test connection
                 char probe_buffer[1];
                 ssize_t probe_result = send(comm->connection->socket_fd, probe_buffer, 0, MSG_NOSIGNAL);
                 
                 if (probe_result < 0 && errno != EAGAIN && errno != EWOULDBLOCK) {
                     // Connection is likely broken
                     printf("Connection appears broken after multiple empty reads\n");
                     comm->connection->connected = 0;
                     break;
                 }
                 
                 // Reset counter after connection check
                 consecutive_empty_reads = 0;
             }
         } else {
             // Actual error occurred in read
             printf("Read error, terminating communication thread\n");
             break;
         }
         
         // Small delay to prevent CPU hogging on busy loop
         usleep(50000);  // 50ms delay
     }
     
     printf("Communication thread exiting\n");
     return NULL;
 }
 
 /**
  * @brief Start the communication thread for message reception
  * 
  * Sets the running flag and creates a background thread that will
  * continuously monitor the connection for incoming messages.
  * 
  * @param comm Pointer to the Communication object
  * 
  * @note The thread will run until stop() is called
  * @note Only one thread per Communication object is supported
  */
 static void Communication_run(Communication* comm) {
     comm->running = 1;
     pthread_create(&comm->thread, NULL, communication_thread_function, comm);
 }
 
 /**
  * @brief Stop the communication thread and wait for termination
  * 
  * Sets the running flag to 0, waits for the background thread to terminate,
  * and resets the thread handle.
  * 
  * @param comm Pointer to the Communication object
  * 
  * @note This function blocks until the thread terminates
  * @note It's safe to call this function multiple times
  */
 static void Communication_stop(Communication* comm) {
     comm->running = 0;
     if (comm->thread) {
         pthread_join(comm->thread, NULL);
         comm->thread = 0;
     }
 }
 
 /**
  * @brief Set the message handler callback function
  * 
  * Assigns a callback function that will be invoked whenever a message
  * is received and successfully decoded.
  * 
  * @param comm Pointer to the Communication object
  * @param handler Message handler callback function
  * 
  * @note The handler can be set to NULL to disable message processing
  * @note The handler is called from the communication thread context
  */
 static void Communication_setMessageHandler(Communication* comm, MessageHandler handler) {
     comm->messageHandler = handler;
 }
 
 /**
  * @brief Create a new Communication object
  * 
  * Allocates memory for a new Communication structure and initializes it
  * with the provided Connection and Protocol objects. All method pointers
  * are assigned for object-oriented usage.
  * 
  * @param connection Pointer to an initialized Connection object
  * @param protocol Pointer to an initialized Protocol object
  * @return Communication* Pointer to the newly created Communication object, or NULL on failure
  * 
  * @note The Connection and Protocol objects are stored by reference, not copied
  * @note The communication thread is not started automatically
  * @note All fields are initialized to safe default values
  */
 Communication* Communication_create(Connection* connection, Protocol* protocol) {
     Communication* comm = (Communication*)malloc(sizeof(Communication));
     if (comm) {
         comm->connection = connection;
         comm->protocol = protocol;
         comm->running = 0;
         comm->thread = 0;
         comm->messageHandler = NULL;
         
         // Assign method pointers
         comm->comX = Communication_comX;
         comm->comY = Communication_comY;
         comm->run = Communication_run;
         comm->stop = Communication_stop;
         comm->setMessageHandler = Communication_setMessageHandler;
     }
     return comm;
 }
 
 /**
  * @brief Destroy a Communication object and free its resources
  * 
  * Stops the communication thread if it's running, waits for termination,
  * and frees the allocated memory. The associated Connection and Protocol
  * objects are not destroyed.
  * 
  * @param comm Pointer to the Communication object to destroy
  * 
  * @note It's safe to pass NULL to this function
  * @note The Connection and Protocol objects are not freed as they may be shared
  * @note If the thread is running, it will be stopped before cleanup
  */
 void Communication_destroy(Communication* comm) {
     if (comm) {
         // Stop the communication thread
         if (comm->running) {
             comm->stop(comm);
         }
         
         // Don't destroy connection and protocol here as they might be shared
         free(comm);
     }
 }